tests/unit-tests/test-class-admin.php

```diff
@@ -0,0 +1,179 @@
+	 */
+	private function duplicate_course_with_lessons_setup( $qty_lessons ) {
+		// Mock the safe_redirect method
+		Sensei()->admin = $this->getMockBuilder( 'WooThemes_Sensei_Admin' )
+			->setMethods( [ 'safe_redirect' ] )
+			->getMock();
+
+		Sensei()->admin->expects( $this->once() )
+			->method( 'safe_redirect' );
+
+		// Create and set current user as teacher
+		$admin_user = $this->factory->user->create_and_get(
+			array(
+				'user_login' => 'admin_user',
+				'user_pass'  => null,
+				'role'       => 'teacher',
+			)
+		);
+		wp_set_current_user( $admin_user->ID );
+
+		// Create one course and add some lessons
+		$course_id   = $this->factory->course->create();
+		$lessons_ids = $this->factory->lesson->create_many( $qty_lessons );
+
+		foreach ( $lessons_ids as $lesson_id ) {
+			add_post_meta( $lesson_id, '_lesson_course', $course_id );
+		}
+
+		// Create nonce
+		$duplicate_nonce = wp_create_nonce( 'duplicate_course_with_lessons_' . $course_id );
+
+		// Set request and get params
+		$_REQUEST['_wpnonce'] = $duplicate_nonce;
+		$_GET['post']         = $course_id;
+
+		return array(
+			'course_id'   => $course_id,
+			'lessons_ids' => $lessons_ids,
+		);
+	}
+
+	/**
+	 * Test duplicate courses with lessons.
+	 *
+	 * @covers WooThemes_Sensei_Admin::duplicate_course_with_lessons_action
+	 *
+	 * @return void
+	 */
+	public function testDuplicateCourseWithLessons() {
+		$this->assertTrue(
+			method_exists( 'WooThemes_Sensei_Admin', 'duplicate_course_with_lessons_action' ),
+			'The admin class function `duplicate_course_with_lessons_action` does not exist '
+		);
+
+		$qty_lessons = 2;
+		$duplication = $this->duplicate_course_with_lessons_setup( $qty_lessons );
+		$course_id   = $duplication['course_id'];
+		$lessons_ids = $duplication['lessons_ids'];
+
+		// Runs the duplication
+		Sensei()->admin->duplicate_course_with_lessons_action();
+
+		// Get the duplicated course
+		$duplicated_course_args = array(
+			'post_type'   => 'course',
+			'meta_key'    => '_duplicate', // phpcs:ignore Slow query ok.
+			'meta_value'  => $course_id, // phpcs:ignore Slow query ok.
+			'post_status' => [ 'draft' ],
+		);
+		$duplicated_courses     = get_posts( $duplicated_course_args );
+		$duplicated_course_id   = $duplicated_courses[0]->ID;
+
+		// Get duplicated lessons
+		$duplicated_lesson_args = array(
+			'post_type'   => 'lesson',
+			'meta_key'    => '_lesson_course', // phpcs:ignore Slow query ok.
+			'meta_value'  => $duplicated_course_id, // phpcs:ignore Slow query ok.
+			'post_status' => [ 'draft' ],
+		);
+		$duplicated_lessons     = get_posts( $duplicated_lesson_args );
+
+		// Lessons assertion
+		$this->assertCount( $qty_lessons, $duplicated_lessons, 'The number of duplicated lessons should be the same as the original.' );
+	}
+
+	/**
+	 * Test duplicate courses with lessons with prerequisite.
+	 *
+	 * @covers WooThemes_Sensei_Admin::duplicate_course_with_lessons_action
```

Comment:
[`@covers`](https://phpunit.readthedocs.io/en/8.4/annotations.html#covers) is used to list the function(s) that we are ultimately writing this test case for. In this case, I believe those are `update_lesson_prerequisite_ids` and `get_prerequisite_update_object`.

[`@uses`](https://phpunit.readthedocs.io/en/8.4/annotations.html#uses) is used to list functions that are called, but are not the functions we're really trying to test. I think `duplicate_course_with_lessons_action` and `duplicate_post` (perhaps others) are examples of this.

Reply:
It's done here: https://github.com/Automattic/sensei/pull/2821/commits/7ff24e12327fef0b2620e50bcd6b8fcda56a4e74

Could you please check if it's in the correct standards?

I think external methods, like `add_post_meta`, `wp_slash`, `get_post_custom`, `get_object_taxonomies`, etc. We don't need to add to `@uses`, correct?

Reply:
Yes, that looks good. I would say `@uses` is a nice to have, and that `@covers` is the more important of the two. 🙂